OpenCL/m17010-pure.cl

```diff
@@ -0,0 +1,444 @@
+    decrypted_data[i + 3] ^= encrypted_data[i + 3];
+
+    // Note: Not necessary if you are only decrypting a single block!
+    essiv[0] = encrypted_data[i + 0];
+    essiv[1] = encrypted_data[i + 1];
+    essiv[2] = encrypted_data[i + 2];
+    essiv[3] = encrypted_data[i + 3];
+  }
+}
+
+DECLSPEC int check_decoded_data (u32 *decoded_data, const u32 decoded_data_size)
+{
+  // Check the SHA-1 of the decrypted data which is stored at the end of the decrypted data
+  const u32 sha1_byte_off = (decoded_data_size - 20);
+  const u32 sha1_u32_off = sha1_byte_off / 4;
+
+  u32 expected_sha1[5];
+  expected_sha1[0] = hc_bytealign_be_S (decoded_data[sha1_u32_off + 1], decoded_data[sha1_u32_off + 0], sha1_byte_off);
+  expected_sha1[1] = hc_bytealign_be_S (decoded_data[sha1_u32_off + 2], decoded_data[sha1_u32_off + 1], sha1_byte_off);
+  expected_sha1[2] = hc_bytealign_be_S (decoded_data[sha1_u32_off + 3], decoded_data[sha1_u32_off + 2], sha1_byte_off);
+  expected_sha1[3] = hc_bytealign_be_S (decoded_data[sha1_u32_off + 4], decoded_data[sha1_u32_off + 3], sha1_byte_off);
+  expected_sha1[4] = hc_bytealign_be_S (decoded_data[sha1_u32_off + 5], decoded_data[sha1_u32_off + 4], sha1_byte_off);
+
+  memzero_be_S (decoded_data, sha1_byte_off, 384 * sizeof(u32));
+
+  sha1_ctx_t ctx;
+
+  sha1_init (&ctx);
+
+  sha1_update_swap (&ctx, decoded_data, sha1_byte_off);
+
+  sha1_final (&ctx);
+
+  return (expected_sha1[0] == hc_swap32_S (ctx.h[0]))
+      && (expected_sha1[1] == hc_swap32_S (ctx.h[1]))
+      && (expected_sha1[2] == hc_swap32_S (ctx.h[2]))
+      && (expected_sha1[3] == hc_swap32_S (ctx.h[3]))
+      && (expected_sha1[4] == hc_swap32_S (ctx.h[4]));
+}
+
+KERNEL_FQ void m17010_init (KERN_ATTR_TMPS_ESALT (gpg_tmp_t, gpg_t))
+{
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  const u32 pw_len = pws[gid].pw_len;
+  const u32 salted_pw_len = (salt_bufs[SALT_POS].salt_len + pw_len);
+
+  u32 salted_pw_block[80];
+
+  // concatenate salt and password -- the salt is always 8 bytes
+  salted_pw_block[0] = salt_bufs[SALT_POS].salt_buf[0];
+  salted_pw_block[1] = salt_bufs[SALT_POS].salt_buf[1];
+
+  for (u32 idx = 0; idx < 64; idx++) salted_pw_block[idx + 2] = pws[gid].i[idx];
+
+  // zero remainder of buffer
+  for (u32 idx = 66; idx < 80; idx++) salted_pw_block[idx] = 0;
+
+  // create a number of copies for efficiency
+  const u32 copies = 80 * sizeof(u32) / salted_pw_len;
+  for (u32 idx = 1; idx < copies; idx++)
+  {
+    memcat_be_S (salted_pw_block, idx * salted_pw_len, salted_pw_block, salted_pw_len);
+  }
+
+  for (u32 idx = 0; idx < 80; idx++) tmps[gid].salted_pw_block[idx] = salted_pw_block[idx];
+
+  tmps[gid].salted_pw_block_len = (copies * salted_pw_len);
+}
+
+KERNEL_FQ void m17010_loop_prepare (KERN_ATTR_TMPS_ESALT (gpg_tmp_t, gpg_t))
+{
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  /**
+   * context save
+   */
+
+  sha1_ctx_t ctx;
+
+  sha1_init (&ctx);
```

Comment:
If you set the initialization h[10] to _init() you can save a lot of branches and complicated code here. Afterwards, the full kernel should look like this:

```
KERNEL_FQ void m17010_loop_prepare (KERN_ATTR_TMPS_ESALT (gpg_tmp_t, gpg_t))
{
  const u64 gid = get_global_id (0);

  if (gid >= gid_max) return;

  tmps[gid].w0[0] = 0;
  tmps[gid].w0[1] = 0;
  tmps[gid].w0[2] = 0;
  tmps[gid].w0[3] = 0;
  tmps[gid].w1[0] = 0;
  tmps[gid].w1[1] = 0;
  tmps[gid].w1[2] = 0;
  tmps[gid].w1[3] = 0;
  tmps[gid].w2[0] = 0;
  tmps[gid].w2[1] = 0;
  tmps[gid].w2[2] = 0;
  tmps[gid].w2[3] = 0;
  tmps[gid].w3[0] = 0;
  tmps[gid].w3[1] = 0;
  tmps[gid].w3[2] = 0;
  tmps[gid].w3[3] = 0;

  tmps[gid].len = salt_repeat;
}
```